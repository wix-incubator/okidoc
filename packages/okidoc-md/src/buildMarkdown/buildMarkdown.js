import remark from 'remark';
import buildMarkdownAST from './buildMarkdownAST';

function buildMarkdown(comments, args) {
  return buildMarkdownAST(comments, args).then(ast => remark().stringify(ast));
}

export default buildMarkdown;
