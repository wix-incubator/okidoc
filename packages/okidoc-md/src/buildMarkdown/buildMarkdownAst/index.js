import buildMarkdownAST from './buildMarkdownAST';

export default buildMarkdownAST;
