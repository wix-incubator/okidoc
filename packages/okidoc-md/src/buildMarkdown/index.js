import buildMarkdown from './buildMarkdown';

export default buildMarkdown;
