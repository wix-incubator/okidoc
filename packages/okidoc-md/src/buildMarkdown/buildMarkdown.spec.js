import documentation from 'documentation';
import buildMarkdown from './';

function getMarkdown(documentationSource, args) {
  return documentation
    .build([{ source: documentationSource }], { shallow: true })
    .then(comments => buildMarkdown(comments, args));
}

describe('buildMarkdown', () => {
  describe('for class', () => {
    it('should render markdown for class with constructor', async () => {
      const documentationSource = `
      /**
      * Component description
      * 
      * @example
      * new Component({visible: true})
      */
      class Component {
        /**
        * creates new Component
        */ 
        constructor(config: any) {}
      
        /**
        * show
        */
        show() {}
      }
    `;
      const markdown = await getMarkdown(documentationSource);

      expect(markdown).toMatchSnapshot();
    });
  });

  describe('for API class', () => {
    const title = 'API Methods';

    it('should render markdown for getters/setters', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * \`isVisible\` getter comment
        */
        get isVisible(): boolean {
            return true;
        }
        
        /**
        * \`width\` setter comment
        */
        set width(value: number) {
            console.log(value)
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods default value', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param [x=''] - \`x\` description 
        * 
        */
        myMethod(x: string) {}
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with @returns comment', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param x - \`x\` description 
        * 
        * @returns returns comment
        */
        myMethod(x: string): string {
          return x;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with UnionType with literals as @param type', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param y - \`y\` comment
        */
        myMethod(y: 'bla' | 'not bla' | 7 | null) {
          return y;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with UnionType with NOT literal as @param type', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param y - \`y\` comment
        */
        myMethod(y: number | null) {
          return y;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with @example end @note', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param x - \`x\` description. see http://example.com
        * 
        * @note
        * \`myMethod\` note 1
        *
        * @example <caption>test example caption</caption>
        * console.log('example')
        *
        * @example
        * console.log('example')
        *
        * @note
        * \`myMethod\` note 2
        * Read more about [myMethod](/my-method)
        *
        */
        myMethod(x: string): string {
          return x;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with @deprecated', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * show component 
        * @deprecated use method \`show()\`
        */
        doShow() {}
        
        /**
        * show component
        */
        show() {}
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with `...rest` as param', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        */
        myMethod(...arr: string[]): string[] {
          return arr;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with `optional` as param', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        */
        myMethod(x: ?number): number {
          return number;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    // pending resolution: https://github.com/documentationjs/documentation/issues/1267
    it.skip('should render markdown for interface with optional property', async () => {
      const documentationSource = `
      /** MyResult interface */
      interface MyResult {
        /** */
        a?: string;
      }
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    // todo
    it.skip('should render markdown for interface with method', async () => {
      const documentationSource = `
        /** */
        interface MyResult {
          /** */
          foo(): string;
        }
        
        /** */
        function bar(): MyResult {}
      `;

      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for interface with method as arrow function', async () => {
      const documentationSource = `
        /** */
        interface MyResult {
          /** */
          foo: () => string;
        }
        
        /** */
        function bar(): MyResult {}
      `;

      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods interface as return type', async () => {
      const documentationSource = `
     /**
      * MyResult interface
      */
      interface MyResult {
        /** @property a - \`a\` description */
        a: string;
      }

      /** Example class jsdoc */
      class API {
        /**
        * myMethod comment
        * @param y - \`y\` comment
        */
        myMethod(y: 'bla' | 'not bla'): MyResult {
          return y;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown sections in valid order', async () => {
      const documentationSource = `
      /** Example class jsdoc */
      class API {
        /**
        * myMethod1 comment
        * @param x - \`x\` description
        * 
        * @example <caption>test example caption</caption>
        * console.log('example')
        * @note
        * \`myMethod1\` note 
        * Read more about [myMethod](/my-method)
        *
        * @returns returns comment
        */
        myMethod1(x: string): string {
          return x;
        }
        
        /**
        * myMethod2 comment
        * @param y - y comment
        * 
        * @example
        * console.log('example')
        */
        myMethod2(y: 'bla' | 'not bla'): string {
          return y;
        }
      }
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });
  });

  describe('for functions', async () => {
    const title = 'Functions';

    it('should render valid markdown', async () => {
      const documentationSource = `
      /**
      * MyFuncResult interface
      */
      interface MyFuncResult {
        /** @property a - \`a\` description */
        a: string;
        /** @property b - \`b\` description */
        b: number;
      }

      /**
      * myFunc1 description
      * @param x - \`x\` description 
      */
      function myFunc1(x: number): number {}

      /**
      * myFunc2 description
      * @param x - \`x\` description 
      * @param y - \`y\` description
      *
      * @example <caption>MyFunc example</caption>
      * 
      * console.log(myFunc(3, 2))
      * // {a: '3', b: 2}
      */
      function myFunc2(x: number, y:number): MyFuncResult {}
    `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render valid markdown with function as argument', async () => {
      const documentationSource = `
        /**
        * @param name - event name 
        * @param fn - event handler 
        */
        function myFunc(name: string, fn: (name: string, data: Object) => any) {}
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render valid markdown with UnionType in TypeApplication as argument', async () => {
      const documentationSource = `
        /**
        * @param fn - event handler 
        */
        function myFunc(fn: (value: string | null) => void) {}
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render valid markdown with function as return type', async () => {
      const documentationSource = `
        /**
        * @returns myFunc result function
        */
        function myFunc(): (flag: boolean) => void {}
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render valid markdown with function in return type interface', async () => {
      const documentationSource = `
        /** */
        interface MyEntity {
          /** */
          id: string;
          /** */
          title: string;
          /** */
          updateTitle: (id: string, title: string) => Promise<any>;
          /** @property destroy - destroy entity */
          destroy: () => void;
        }
      
        /**
        * create new entity
        * @param title - entity title
        * @returns new entity
        */
        function create(title: string): MyEntity {}
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });
  });

  describe('with ApplicationType as return type', () => {
    const title = 'API Methods';

    it('should render markdown for methods with Promise as @returns type', async () => {
      const documentationSource = `
      /**
      * @returns returns comment
      */
      function myFn(): Promise<string> {
        return x;
      }
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with Promise & interface as @returns type', async () => {
      const documentationSource = `
      /** */
      interface IResult {
        /** */
        x: number;
        /** */
        y: number;
      }

      /**
      * @returns returns comment
      */
      function myFn(): Promise<IResult> {
        return {x: 1, y: 1};
      }
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with Array as @returns type', async () => {
      const documentationSource = `
      /**
      * @returns returns comment
      */
      function myFn(): string[] {
        return [''];
      }
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });

    it('should render markdown for methods with Array & interface as @returns type', async () => {
      const documentationSource = `
      /** */
      interface IResult {
        /** */
        x: number;
        /** */
        y: number;
      }

      /**
      * @returns returns comment
      */
      function myFn(): IResult[] {
        return [{x: 1, y: 1}];
      }
      `;
      const markdown = await getMarkdown(documentationSource, { title });

      expect(markdown).toMatchSnapshot();
    });
  });
});
