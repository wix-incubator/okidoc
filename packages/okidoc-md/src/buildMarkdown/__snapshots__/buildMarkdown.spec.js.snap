// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMarkdown for API class should render markdown for getters/setters 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## isVisible

\`isVisible\` getter comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">boolean</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

## width

\`width\` setter comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>value</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods default value 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
          <div class=\\"optional\\">(Optional)</div>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>x</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods interface as return type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <p><code>y</code> comment</p>
            Possible values are <code>\\"bla\\"</code>, <code>\\"not bla\\"</code>.
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>a</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>a</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with @deprecated 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## doShow()

**DEPRECATED!**

use method \`show()\`

show component

## show()

show component
"
`;

exports[`buildMarkdown for API class should render markdown for methods with @example end @note 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

> \`myMethod\` note 1

> test example caption

\`\`\`javascript
console.log('example')
\`\`\`

\`\`\`javascript
console.log('example')
\`\`\`

> \`myMethod\` note 2
> Read more about [myMethod](/my-method)

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>x</code> description. see <a href=\\"http://example.com\\">http://example.com</a></p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">string</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with @returns comment 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>x</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p>returns comment</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with \`...rest\` as param 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>arr</code>
        </td>
        <td>
            <div class=\\"type\\">...Array&#x3C;string></div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">Array&#x3C;string></div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with \`optional\` as param 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number?</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with UnionType with NOT literal as @param type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <div class=\\"type\\">(number | null)</div>
            <p><code>y</code> comment</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown for methods with UnionType with literals as @param type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod()

myMethod comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <p><code>y</code> comment</p>
            Possible values are <code>\\"bla\\"</code>, <code>\\"not bla\\"</code>, <code>7</code>, null.
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for API class should render markdown sections in valid order 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myMethod1()

> test example caption

\`\`\`javascript
console.log('example')
\`\`\`

> \`myMethod1\` note 
> Read more about [myMethod](/my-method)

myMethod1 comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>x</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p>returns comment</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

## myMethod2()

\`\`\`javascript
console.log('example')
\`\`\`

myMethod2 comment

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <p>y comment</p>
            Possible values are <code>\\"bla\\"</code>, <code>\\"not bla\\"</code>.
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">string</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for class should render markdown for class with constructor 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Component

\`\`\`javascript
new Component({visible: true})
\`\`\`

Component description

\`new Component()\`

creates new Component

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>config</code>
        </td>
        <td>
            <div class=\\"type\\">any</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

## show()

show
"
`;

exports[`buildMarkdown for functions should render valid markdown 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Functions

## myFunc1()

myFunc1 description

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
            <p><code>x</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

## myFunc2()

> MyFunc example

\`\`\`javascript
console.log(myFunc(3, 2))
// {a: '3', b: 2}
\`\`\`

myFunc2 description

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
            <p><code>x</code> description</p>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
            <p><code>y</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>a</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p><code>a</code> description</p>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>b</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
            <p><code>b</code> description</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for functions should render valid markdown with UnionType in TypeApplication as argument 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Functions

## myFunc()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>fn</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(<code>value</code>: (string | null)): void</div>
            <p>event handler</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for functions should render valid markdown with function as argument 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Functions

## myFunc()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>name</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p>event name</p>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>fn</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(<code>name</code>: string, <code>data</code>: Object): any</div>
            <p>event handler</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for functions should render valid markdown with function as return type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Functions

## myFunc()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"function-type\\">function(<code>flag</code>: boolean): void</div>
            <p>myFunc result function</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown for functions should render valid markdown with function in return type interface 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Functions

## create()

create new entity

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">ARGUMENTS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>title</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
            <p>entity title</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>id</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>title</code>
        </td>
        <td>
            <div class=\\"type\\">string</div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>updateTitle</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(<code>id</code>: string, <code>title</code>: string): Promise&#x3C;any></div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>destroy</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(): void</div>
            <p>destroy entity</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown with ApplicationType as return type should render markdown for methods with Array & interface as @returns type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myFn()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th><div class=\\"type\\">Array&#x3C;IResult></div></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown with ApplicationType as return type should render markdown for methods with Array as @returns type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myFn()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">Array&#x3C;string></div>
            <p>returns comment</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown with ApplicationType as return type should render markdown for methods with Promise & interface as @returns type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myFn()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th><div class=\\"type\\">Promise&#x3C;IResult></div></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>y</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildMarkdown with ApplicationType as return type should render markdown for methods with Promise as @returns type 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# API Methods

## myFn()

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
        </td>
        <td>
            <div class=\\"type\\">Promise&#x3C;string></div>
            <p>returns comment</p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;
